core/src/nodes/listeners.rs

```diff
@@ -181,15 +182,15 @@ where
     /// Provides an API similar to `Stream`, except that it cannot error.
     pub fn poll(&mut self) -> Async<ListenersEvent<TTrans>> {
         // We remove each element from `listeners` one by one and add them back.
-        for n in (0..self.listeners.len()).rev() {
-            let mut listener = self.listeners.swap_remove(n);
+        for _n in 0..self.listeners.len() {
+            let mut listener = self.listeners.pop_back().expect("There is at least one Listener if we got here; qed");
```

Comment:
Personally I would have tried to avoid unwrapping:

```rust
let mut remaining = self.listeners.len();
while let Some(listener) = self.listers.pop_back() {
    match listener.listener.poll() {
        Ok(Async::NotReady) => {
            self.listeners.push_front();
            remaining -= 1;
            if remaining == 0 { break }
        }
        ...
    }
}
```